crates/oxc_sourcemap/src/encode.rs

```diff
@@ -26,37 +28,38 @@ pub fn encode(sourcemap: &SourceMap) -> JSONSourceMap {
 // Here using `serde_json::to_string` to serialization `names/source_contents/sources`.
 // It will escape the string to avoid invalid JSON string.
 pub fn encode_to_string(sourcemap: &SourceMap) -> Result<String> {
-    let mut buf = String::new();
-    buf.push_str("{\"version\":3,");
+    // Max length of `contents` is 18.
+    let mut contents = PreAllocatedString::new(18);
+    contents.push("{\"version\":3,".into());
     if let Some(file) = sourcemap.get_file() {
-        buf.push_str("\"file\":\"");
-        buf.push_str(file);
-        buf.push_str("\",");
+        contents.push("\"file\":\"".into());
+        contents.push(file.into());
+        contents.push("\",".into());
     }
     if let Some(source_root) = sourcemap.get_source_root() {
-        buf.push_str("\"sourceRoot\":\"");
-        buf.push_str(source_root);
-        buf.push_str("\",");
+        contents.push("\"sourceRoot\":\"".into());
+        contents.push(source_root.into());
+        contents.push("\",".into());
     }
-    buf.push_str("\"names\":[");
+    contents.push("\"names\":[".into());
     let names = sourcemap
         .names
         .iter()
         .map(|x| serde_json::to_string(x.as_ref()))
         .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
         .map_err(Error::from)?;
-    buf.push_str(&names.join(","));
-    buf.push_str("],\"sources\":[");
+    contents.push(Cow::Owned(names.join(",")));
```

Comment:
Rather than `join`ing here, how about allocating slots in `contents` for each name, and pushing them individually?

```rs
// NB 17 not 16 to handle case where `sourcemap.names.len() == 0`
let mut contents = PreAllocatedString::new(17 + sourcemap.names.len() * 2);
// ...
contents.push("\"names\":[".into());
for name in &sourcemap.names {
    let encoded = serde_json::to_string(name.as_ref()).map_err(Error::from)?;
    contents.push(Cow::Owned(encoded));
    contents.push(Cow::Borrowed(","));
}
contents.pop();
```

Could do same for all the other strings which are currently `join`ed. It should have biggest effect with `source_contents`, as will avoid copying those long strings twice.